### PR TITLE
Fix #6301 to re-sort the beans

### LINKED ISSUE #6301: Bean autowiring is non-deterministic when multiple beans exist with no @Priority
### Bug description

Relates to #5464. The code changes for that feature assume that if there are multiple beans for a given type, that they will have an `@Priority` assigned, so that the desired bean is returned from registry lookups.

IMO this is only reliable for beans declared from user code. The majority of beans created by Quarkus extensions do not have any `@Priority` (AFAIK). E.g like `DataSource` beans.

Thus, when Camel attempts to do autowiring, it will potentially inject a 'random' bean into components, since there is no `@Priority` enforced.

### PR DESCRIPTION
**Pull request by zhfeng**
Fix 
- #6301

### CODE REVIEW

**Review by oscerd — APPROVED** [review]

File: extensions-core/core/runtime/src/main/java/org/apache/camel/quarkus/core/RuntimeBeanRepository.java [hunk 1]
```diff
@@ -163,8 +165,13 @@ public <T> T findSingleByType(Class<T> type) {
             } else {
                 handles = container.listAll(type);
             }
+
             if (handles.size() > 0) {
-                return handles.get(0).get();
+                if (handles.stream().filter(h -> !h.getBean().getKind().equals(InjectableBean.Kind.SYNTHETIC)).count() > 0) {
```
Comment (jamesnetherton) [inline extensions-core/core/runtime/src/main/java/org/apache/camel/quarkus/core/RuntimeBeanRepository.java:170]: What's the significance of the `SYNTHETIC` beans kind where we can just take the first one? Is that safe if multiple results are returned?

File: extensions-core/core/runtime/src/main/java/org/apache/camel/quarkus/core/RuntimeBeanRepository.java [hunk 2]
```diff
@@ -163,8 +165,13 @@ public <T> T findSingleByType(Class<T> type) {
...
+                    return handles.get(0).get();
+                } else {
+                    return container.select(type, Default.Literal.INSTANCE).get();
```
Comment (jamesnetherton) [inline extensions-core/core/runtime/src/main/java/org/apache/camel/quarkus/core/RuntimeBeanRepository.java:173]: Is it safe to call `get()` if no default bean exists? IIRC it'll throw an exception if not.

**Review by jamesnetherton — COMMENTED** [review]
Thanks @zhfeng. I left a couple of question comments.

File: extensions-core/core/runtime/src/main/java/org/apache/camel/quarkus/core/RuntimeBeanRepository.java (hunk 1 above)
Comment (zhfeng) [reply extensions-core/core/runtime/src/main/java/org/apache/camel/quarkus/core/RuntimeBeanRepository.java:170]: Synthetic beans are usually provided by extensions. I'm thinking another way with resorting the handels?

The orignal `handles` is sorted by `priority` and when `priority` is equal, the position is random. How about to introduce a new `Comparator` if `priority` is the same, the default bean is win?

Comment (jamesnetherton) [reply extensions-core/core/runtime/src/main/java/org/apache/camel/quarkus/core/RuntimeBeanRepository.java:170]: > introduce a new Comparator if priority is the same, the default bean is win?

Yeah, sounds ok.

Probably faster than the alternative, which would be to collect all priorities in a set and then check the size.

Comment (zhfeng) [reply extensions-core/core/runtime/src/main/java/org/apache/camel/quarkus/core/RuntimeBeanRepository.java]: OK, I updated with the new `Comparator`.

File: extensions-core/core/runtime/src/main/java/org/apache/camel/quarkus/core/RuntimeBeanRepository.java (hunk 2 above)
Comment (zhfeng) [reply extensions-core/core/runtime/src/main/java/org/apache/camel/quarkus/core/RuntimeBeanRepository.java]: Remove it.

**Review by jamesnetherton — APPROVED** [review]

**Pull request merged**